--- transolver/Airfoil-Design-AirfRANS/main.py
+++ linearno/AirfRANS/main.py
@@ -19,11 +19,16 @@
                     help='If you want to compute the score of the models on the associated test set. (default: 0)',
                     default=0, type=int)
 parser.add_argument('--my_path',
-                    default='/data/path', type=str)
+                    default='./datas/Dataset', type=str)
 parser.add_argument('--save_path',
                     default='metrics', type=str)
+parser.add_argument('--save_name',
+                    default='default', type=str)
+parser.add_argument('--debug',
+                    default=1, type=int)
 args = parser.parse_args()
-
+args.save_path=osp.join(args.save_path,args.save_name)
+print(args)
 with open(args.my_path + '/manifest.json', 'r') as f:
     manifest = json.load(f)
 
@@ -32,6 +37,10 @@
 n = int(.1 * len(manifest_train))
 train_dataset = manifest_train[:-n]
 val_dataset = manifest_train[-n:]
+if bool(args.debug):
+    print("debuging")
+    train_dataset=train_dataset[:2]
+    val_dataset=val_dataset[:1]
 print("start load data")
 train_dataset, coef_norm = Dataset(train_dataset, norm=True, sample=None, my_path=args.my_path)
 val_dataset = Dataset(val_dataset, sample=None, coef_norm=coef_norm, my_path=args.my_path)
@@ -51,11 +60,9 @@
 
 models = []
 for i in range(args.nmodel):
-
-    if args.model == 'Transolver':
-        from models.Transolver import Transolver
-
-        model = Transolver(n_hidden=256,
+    if args.model == 'LinearAttentionNeuralOperator':
+        from models.LinearAttnNeuralOperator import LinearAttentionNeuralOperator
+        model= LinearAttentionNeuralOperator(n_hidden=256,
                            n_layers=8,
                            space_dim=7,
                            fun_dim=0,
@@ -63,7 +70,8 @@
                            mlp_ratio=2,
                            out_dim=4,
                            slice_num=32,
-                           unified_pos=1).cuda()
+                           unified_pos=1,
+                           linear=True).cuda()
     else:
         encoder = MLP(hparams['encoder'], batch_norm=False)
         decoder = MLP(hparams['decoder'], batch_norm=False)
@@ -89,26 +97,30 @@
 
     log_path = osp.join(args.save_path, args.task, args.model)  # path where you want to save log and figures
     print('start training')
+    print(model)
+    print(f'Number of parameters: {sum(p.numel() for p in model.parameters() if p.requires_grad)}')
+    
     model = train.main(device, train_dataset, val_dataset, model, hparams, log_path,
                        criterion='MSE_weighted', val_iter=10, reg=args.weight, name_mod=args.model, val_sample=True)
     print('end training')
     models.append(model)
-torch.save(models, osp.join(args.save_path, args.task, args.model, args.model))
+torch.save(models, osp.join(args.save_path,args.task, args.model, args.model))
 
 if bool(args.score):
     print('start score')
     s = args.task + '_test' if args.task != 'scarce' else 'full_test'
-    coefs = metrics.Results_test(device, [models], [hparams], coef_norm, args.my_path, path_out='scores', n_test=3,
+    path_out=osp.join('scores',args.save_name)
+    coefs = metrics.Results_test(device, [models], [hparams], coef_norm, args.my_path, path_out=path_out, n_test=3,
                                  criterion='MSE', s=s)
     # models can be a stack of the same model (for example MLP) on the task s, if you have another stack of another model (for example GraphSAGE)
     # you can put in model argument [models_MLP, models_GraphSAGE] and it will output the results for both models (mean and std) in an ordered array.
-    np.save(osp.join('scores', args.task, 'true_coefs'), coefs[0])
-    np.save(osp.join('scores', args.task, 'pred_coefs_mean'), coefs[1])
-    np.save(osp.join('scores', args.task, 'pred_coefs_std'), coefs[2])
+    np.save(osp.join(path_out, args.task, 'true_coefs'), coefs[0])
+    np.save(osp.join(path_out, args.task, 'pred_coefs_mean'), coefs[1])
+    np.save(osp.join(path_out, args.task, 'pred_coefs_std'), coefs[2])
     for n, file in enumerate(coefs[3]):
-        np.save(osp.join('scores', args.task, 'true_surf_coefs_' + str(n)), file)
+        np.save(osp.join(path_out, args.task, 'true_surf_coefs_' + str(n)), file)
     for n, file in enumerate(coefs[4]):
-        np.save(osp.join('scores', args.task, 'surf_coefs_' + str(n)), file)
-    np.save(osp.join('scores', args.task, 'true_bls'), coefs[5])
-    np.save(osp.join('scores', args.task, 'bls'), coefs[6])
+        np.save(osp.join(path_out, args.task, 'surf_coefs_' + str(n)), file)
+    np.save(osp.join(path_out, args.task, 'true_bls'), coefs[5])
+    np.save(osp.join(path_out, args.task, 'bls'), coefs[6])
     print('end score')
--- transolver/Airfoil-Design-AirfRANS/main_evaluation.py
+++ linearno/AirfRANS/main_evaluation.py
@@ -7,7 +7,7 @@
 import numpy as np
 
 parser = argparse.ArgumentParser()
-parser.add_argument('--my_path', default='/data/path', type=str)  # data save path
+parser.add_argument('--my_path', default='./data', type=str)  # data save path
 parser.add_argument('--save_path', default='./', type=str)  # model save path
 args = parser.parse_args()
 
@@ -23,7 +23,7 @@
 data_root_dir = args.my_path
 ckpt_root_dir = args.save_path
 
-tasks = ['full']
+tasks = ['full','aoa','reynolds']
 
 for task in tasks:
     print('Generating results for task ' + task + '...')
@@ -37,28 +37,25 @@
 
     manifest_train = manifest[s_train]
     n = int(.1 * len(manifest_train))
+    # train_dataset = manifest_train[:1]
     train_dataset = manifest_train[:-n]
-
     _, coef_norm = Dataset(train_dataset, norm=True, sample=None, my_path=data_dir)
-
     # Compute the scores on the test set
-
-    model_names = ['Transolver']
+    model_names = ['LinearAttentionNeuralOperator']
     models = []
     hparams = []
-
     for model in model_names:
-        model_path = osp.join(ckpt_root_dir, 'metrics', task, model, model)
+        model_path = osp.join(ckpt_root_dir, 'metrics', model, task, model, "model")
         mod = torch.load(model_path)
         print(mod)
-        mod = [m.to(device) for m in mod]
-        models.append(mod)
+        # mod = [m.to(device) for m in mod]
+        models.append([mod])
 
         with open('params.yaml', 'r') as f:
             hparam = yaml.safe_load(f)[model]
             hparams.append(hparam)
 
-    results_dir = osp.join(ckpt_root_dir, 'scores', task)
+    results_dir = osp.join(ckpt_root_dir, 'scores')
     coefs = metrics.Results_test(device, models, hparams, coef_norm, data_dir, results_dir, n_test=3, criterion='MSE',
                                  s=s)
     # models can be a stack of the same model (for example MLP) on the task s, if you have another stack of another model (for example GraphSAGE)
--- transolver/Airfoil-Design-AirfRANS/params.yaml
+++ linearno/AirfRANS/params.yaml
@@ -14,8 +14,22 @@
 
 Transolver:
   batch_size: 1
-  nb_epochs: 398
+  nb_epochs: 400
   lr: 0.001
+  max_neighbors: 64
+  subsampling: 32000
+  r: 0.05
+LinearAttentionNeuralOperator:
+  batch_size: 1
+  nb_epochs: 400
+  lr: 0.001
+  max_neighbors: 64
+  subsampling: 32000
+  r: 0.05
+LNO:
+  batch_size: 1
+  nb_epochs: 400
+  lr: 0.0001
   max_neighbors: 64
   subsampling: 32000
   r: 0.05
--- transolver/Airfoil-Design-AirfRANS/scripts/Evaluation.sh
+++ linearno/AirfRANS/scripts/Evaluation.sh
@@ -1,3 +1,3 @@
-export CUDA_VISIBLE_DEVICES=4
+export CUDA_VISIBLE_DEVICES=0
 
-python main_evaluation.py --my_path /data/naca/+python main_evaluation.py --my_path ./datas/--- transolver/Airfoil-Design-AirfRANS/train.py
+++ linearno/AirfRANS/train.py
@@ -115,6 +115,8 @@
     def default(self, obj):
         if isinstance(obj, np.ndarray):
             return obj.tolist()
+        elif isinstance(obj,np.float32):
+            return float(obj)
         return json.JSONEncoder.default(self, obj)
 
 
@@ -167,7 +169,8 @@
             data_sampled.y = data_sampled.y[idx]
             data_sampled.surf = data_sampled.surf[idx]
 
-            if name_mod != 'PointNet' and name_mod != 'MLP':
+            if name_mod != 'PointNet' and name_mod != 'MLP' and name_mod != 'LinearAttentionNeuralOperator' :
+                print("计算knn")
                 data_sampled.edge_index = nng.radius_graph(x=data_sampled.pos.to(device), r=hparams['r'], loop=True,
                                                            max_num_neighbors=int(hparams['max_neighbors'])).cpu()
 
@@ -207,7 +210,7 @@
                             data_sampled.y = data_sampled.y[idx]
                             data_sampled.surf = data_sampled.surf[idx]
 
-                            if name_mod != 'PointNet' and name_mod != 'MLP':
+                            if name_mod != 'PointNet' and name_mod != 'MLP' and name_mod != 'LinearAttentionNeuralOperator':
                                 data_sampled.edge_index = nng.radius_graph(x=data_sampled.pos.to(device),
                                                                            r=hparams['r'], loop=True,
                                                                            max_num_neighbors=int(
@@ -329,14 +332,14 @@
                         'nb_parameters': params_model,
                         'time_elapsed': time_elapsed,
                         'hparams': hparams,
-                        'train_loss_surf': str(train_loss_surf_list[-1]),
-                        'train_loss_surf_var': str(loss_surf_var_list[-1]),
-                        'train_loss_vol': str(train_loss_vol_list[-1]),
-                        'train_loss_vol_var': str(loss_vol_var_list[-1]),
-                        'val_loss_surf': str(val_surf_list[-1]),
-                        'val_loss_surf_var': str(val_surf_var_list[-1]),
-                        'val_loss_vol': str(val_vol_list[-1]),
-                        'val_loss_vol_var': str(val_vol_var_list[-1]),
+                        'train_loss_surf': train_loss_surf_list[-1],
+                        'train_loss_surf_var': loss_surf_var_list[-1],
+                        'train_loss_vol': train_loss_vol_list[-1],
+                        'train_loss_vol_var': loss_vol_var_list[-1],
+                        'val_loss_surf': val_surf_list[-1],
+                        'val_loss_surf_var': val_surf_var_list[-1],
+                        'val_loss_vol': val_vol_list[-1],
+                        'val_loss_vol_var': val_vol_var_list[-1],
                     }, f, indent=12, cls=NumpyEncoder
                 )
 
--- transolver/Airfoil-Design-AirfRANS/utils/metrics.py
+++ linearno/AirfRANS/utils/metrics.py
@@ -273,7 +273,15 @@
         manifest = json.load(f)
 
     test_dataset = manifest[s]
+    # idx = []
+    # for i,data in enumerate(test_dataset):
+    #     print(data)
+    #     if data != 'airFoil2D_SST_32.892_-0.128_5.528_3.635_15.155' and data != 'airFoil2D_SST_48.233_10.709_5.735_0.0_13.559'\
+    #             and data != 'airFoil2D_SST_64.173_1.751_1.546_7.336_1.0_18.982':
+    #         continue
+    #     idx.append(i)
     idx = random.sample(range(len(test_dataset)), k=n_test)
+    print(idx)
     idx.sort()
 
     test_dataset_vtk = Dataset(test_dataset, sample=None, coef_norm=coef_norm, my_path=path_in)
@@ -312,6 +320,8 @@
         pred_coef = []
 
         for j, data in enumerate(tqdm(test_loader)):
+
+            
             Uinf, angle = float(test_dataset[j].split('_')[2]), float(test_dataset[j].split('_')[3])
             outs, tim = Infer_test(device, model, hparams, data, coef_norm=coef_norm)
             times.append(tim)
@@ -337,6 +347,7 @@
                 loss_per_var = criterion(out, data.y).mean(dim=0)
                 loss = loss_per_var.mean()
                 loss_surf_var = criterion(out[data.surf, :], data.y[data.surf, :]).mean(dim=0)
+                # print(out[data.surf, :10], data.y[data.surf, :10])
                 loss_vol_var = criterion(out[~data.surf, :], data.y[~data.surf, :]).mean(dim=0)
                 loss_surf = loss_surf_var.mean()
                 loss_vol = loss_vol_var.mean()
@@ -419,6 +430,7 @@
     bls = []
     true_bls = []
     for i in range(len(internals[0])):
+        
         aero_name = test_dataset[idx[i]]
         true_internal = true_internals[i]
         true_airfoil = true_airfoils[i]
--- transolver/Car-Design-ShapeNetCar/dataset/dataset.py
+++ linearno/ShapeNetCar/dataset/dataset.py
@@ -147,6 +147,7 @@
             if not os.path.exists(save_path):
                 continue
             init = np.load(os.path.join(save_path, 'x.npy'))
+           
             target = np.load(os.path.join(save_path, 'y.npy'))
             pos = np.load(os.path.join(save_path, 'pos.npy'))
             surf = np.load(os.path.join(save_path, 'surf.npy'))
--- transolver/Car-Design-ShapeNetCar/dataset/load_dataset.py
+++ linearno/ShapeNetCar/dataset/load_dataset.py
@@ -30,6 +30,7 @@
     print("loading data")
     train_dataset, coef_norm = get_datalist(args.data_dir, trainlst, norm=True, savedir=args.save_dir,
                                             preprocessed=preprocessed)
+    
     val_dataset = get_datalist(args.data_dir, vallst, coef_norm=coef_norm, savedir=args.save_dir,
                                preprocessed=preprocessed)
     print("load data finish")
--- transolver/Car-Design-ShapeNetCar/main.py
+++ linearno/ShapeNetCar/main.py
@@ -5,11 +5,10 @@
 
 from dataset.load_dataset import load_train_val_fold
 from dataset.dataset import GraphDataset
-from models.Transolver import Model
-
+from models.LinearAttnNeuralOperator import LinearAttentionNeuralOperator
 parser = argparse.ArgumentParser()
-parser.add_argument('--data_dir', default='/data/PDE_data/mlcfd_data/training_data')
-parser.add_argument('--save_dir', default='/data/PDE_data/mlcfd_data/preprocessed_data')
+parser.add_argument('--data_dir', default='./datas/mlcfd_data/training_data')
+parser.add_argument('--save_dir', default='./datas/mlcfd_data/preprocessed_data')
 parser.add_argument('--fold_id', default=0, type=int)
 parser.add_argument('--gpu', default=0, type=int)
 parser.add_argument('--val_iter', default=10, type=int)
@@ -19,8 +18,8 @@
 parser.add_argument('--r', default=0.2, type=float)
 parser.add_argument('--weight', default=0.5, type=float)
 parser.add_argument('--lr', default=0.001, type=float)
-parser.add_argument('--batch_size', default=1, type=int)
-parser.add_argument('--nb_epochs', default=200, type=int)
+parser.add_argument('--batch_size', default=1, type=float)
+parser.add_argument('--nb_epochs', default=200, type=float)
 parser.add_argument('--preprocessed', default=1, type=int)
 args = parser.parse_args()
 print(args)
@@ -35,13 +34,11 @@
 train_ds = GraphDataset(train_data, use_cfd_mesh=args.cfd_mesh, r=args.r)
 val_ds = GraphDataset(val_data, use_cfd_mesh=args.cfd_mesh, r=args.r)
 
-if args.cfd_model == 'Transolver':
-    model = Model(n_hidden=256, n_layers=8, space_dim=7,
-                  fun_dim=0,
-                  n_head=8,
-                  mlp_ratio=2, out_dim=4,
-                  slice_num=32,
-                  unified_pos=0).cuda()
+
+if args.cfd_model == 'LinearAttnNeuralOperator':
+    model = LinearAttentionNeuralOperator(space_dim=3,n_layers=8,n_hidden=256,n_head=8,Time_Input=False,
+                                 mlp_ratio=2,fun_dim=4,out_dim=4,key_ratio=1,unified_pos=False,isregular=False).cuda()
+
 
 path = f'metrics/{args.cfd_model}/{args.fold_id}/{args.nb_epochs}_{args.weight}'
 if not os.path.exists(path):
--- transolver/Car-Design-ShapeNetCar/main_evaluation.py
+++ linearno/ShapeNetCar/main_evaluation.py
@@ -12,8 +12,8 @@
 import scipy as sc
 
 parser = argparse.ArgumentParser()
-parser.add_argument('--data_dir', default='/data/PDE_data/mlcfd_data/training_data')
-parser.add_argument('--save_dir', default='/data/PDE_data/mlcfd_data/preprocessed_data')
+parser.add_argument('--data_dir', default='/data/mlcfd_data/training_data')
+parser.add_argument('--save_dir', default='/data/mlcfd_data/preprocessed_data')
 parser.add_argument('--fold_id', default=0, type=int)
 parser.add_argument('--gpu', default=0, type=int)
 parser.add_argument('--cfd_model')
@@ -33,7 +33,8 @@
 val_ds = GraphDataset(val_data, use_cfd_mesh=args.cfd_mesh, r=args.r)
 
 path = f'metrics/{args.cfd_model}/{args.fold_id}/{args.nb_epochs}_{args.weight}'
-model = torch.load(os.path.join(path, f'model_{args.nb_epochs}.pth')).to(device)
+print(path)
+model = torch.load(os.path.join(path, f'model_{args.nb_epochs}.pth'),map_location=device)
 
 test_loader = DataLoader(val_ds, batch_size=1)
 
@@ -66,8 +67,6 @@
             std = torch.tensor(coef_norm[3]).to(device)
             pred_press = out[cfd_data.surf, -1] * std[-1] + mean[-1]
             gt_press = targets[cfd_data.surf, -1] * std[-1] + mean[-1]
-            pred_surf_velo = out[cfd_data.surf, :-1] * std[:-1] + mean[:-1]
-            gt_surf_velo = targets[cfd_data.surf, :-1] * std[:-1] + mean[:-1]
             pred_velo = out[~cfd_data.surf, :-1] * std[:-1] + mean[:-1]
             gt_velo = targets[~cfd_data.surf, :-1] * std[:-1] + mean[:-1]
             out_denorm = out * std + mean
@@ -75,11 +74,11 @@
 
         np.save('./results/' + args.cfd_model + '/' + str(index) + '_pred.npy', out_denorm.detach().cpu().numpy())
         np.save('./results/' + args.cfd_model + '/' + str(index) + '_gt.npy', y_denorm.detach().cpu().numpy())
-
+        
         pred_coef = cal_coefficient(vallst[index].split('/')[1], pred_press[:, None].detach().cpu().numpy(),
-                                    pred_surf_velo.detach().cpu().numpy())
+                                    pred_velo.detach().cpu().numpy())
         gt_coef = cal_coefficient(vallst[index].split('/')[1], gt_press[:, None].detach().cpu().numpy(),
-                                  gt_surf_velo.detach().cpu().numpy())
+                                  gt_velo.detach().cpu().numpy())
 
         gt_coef_list.append(gt_coef)
         pred_coef_list.append(pred_coef)
--- transolver/Car-Design-ShapeNetCar/scripts/Evaluation.sh
+++ linearno/ShapeNetCar/scripts/Evaluation.sh
@@ -1,6 +1,6 @@
-export CUDA_VISIBLE_DEVICES=3
+export CUDA_VISIBLE_DEVICES=0
 
 python main_evaluation.py \
---cfd_model=Transolver \
---data_dir /data/PDE_data/mlcfd_data/training_data \
---save_dir /data/PDE_data/mlcfd_data/preprocessed_data \+--cfd_model=LinearAttnNeuralOperator \
+--data_dir ./datas/mlcfd_data/training_data \
+--save_dir ./datas/mlcfd_data/preprocessed_data \
--- transolver/Car-Design-ShapeNetCar/utils/drag_coefficient.py
+++ linearno/ShapeNetCar/utils/drag_coefficient.py
@@ -146,8 +146,8 @@
 
 ############## calculate coefficient ##############
 def cal_coefficient(file_name, press_surf=None, velo_surf=None):
-    root = '/data/PDE_data/mlcfd_data/training_data'
-    save_path = '/data/PDE_data/mlcfd_data/preprocessed_data/param0/' + file_name
+    root = './datas/mlcfd_data/training_data'
+    save_path = './datas/mlcfd_data/param0/' + file_name
     file_name_press = 'param0/' + file_name + '/quadpress_smpl.vtk'
     file_name_velo = 'param0/' + file_name + '/hexvelo_smpl.vtk'
     file_name_press = os.path.join(root, file_name_press)
--- transolver/PDE-Solving-StandardBenchmark/exp_airfoil.py
+++ linearno/Standard_PDE_Benchmark/exp_airfoil.py
@@ -5,14 +5,13 @@
 import torch
 from tqdm import *
 from utils.testloss import TestLoss
-from model_dict import get_model
-
-parser = argparse.ArgumentParser('Training Transformer')
+from torch.utils.tensorboard import SummaryWriter
+from model.LinearAttnNeuralOperator import LinearAttentionNeuralOperator
+parser = argparse.ArgumentParser('Training')
 
 parser.add_argument('--lr', type=float, default=1e-3)
 parser.add_argument('--epochs', type=int, default=500)
 parser.add_argument('--weight_decay', type=float, default=1e-5)
-parser.add_argument('--model', type=str, default='Transolver_2D')
 parser.add_argument('--n-hidden', type=int, default=64, help='hidden dim')
 parser.add_argument('--n-layers', type=int, default=3, help='layers')
 parser.add_argument('--n-heads', type=int, default=4)
@@ -25,16 +24,21 @@
 parser.add_argument('--dropout', type=float, default=0.0)
 parser.add_argument('--unified_pos', type=int, default=0)
 parser.add_argument('--ref', type=int, default=8)
-parser.add_argument('--slice_num', type=int, default=32)
+parser.add_argument('--key_ratio', type=int, default=4)
 parser.add_argument('--eval', type=int, default=0)
-parser.add_argument('--save_name', type=str, default='airfoil_Transolver')
+parser.add_argument('--save_name', type=str, default='airfoil_LinearAttnNeuralOperator')
 parser.add_argument('--data_path', type=str, default='/data/fno/airfoil/naca')
+parser.add_argument("--model",type=str,default='None')
 args = parser.parse_args()
 eval = args.eval
 save_name = args.save_name
 
 os.environ["CUDA_VISIBLE_DEVICES"] = args.gpu
-
+def set_seed(num):
+    torch.manual_seed(num)
+    torch.cuda.manual_seed_all(num)
+    np.random.seed(num)
+    torch.backends.cudnn.deterministic = True
 
 def count_parameters(model):
     total_params = 0
@@ -45,14 +49,25 @@
     print(f"Total Trainable Params: {total_params}")
     return total_params
 
+def count_parameters1(model):
+    print(f"{'Module':40s} | {'Params':>10s}")
+    print("-" * 55)
+    total = 0
+    for name, param in model.named_parameters():
+        if param.requires_grad:
+            num = param.numel()
+            print(f"{name:40s} | {num:10d}")
+            total += num
+    print("-" * 55)
+    print(f"{'Total Trainable Params':40s} | {total:10d}")
+    return total
+
 
 def main():
+    # set_seed(0)
     INPUT_X = args.data_path + '/NACA_Cylinder_X.npy'
     INPUT_Y = args.data_path + '/NACA_Cylinder_Y.npy'
     OUTPUT_Sigma = args.data_path + '/NACA_Cylinder_Q.npy'
-
-    ntrain = 1000
-    ntest = 200
 
     r1 = args.downsamplex
     r2 = args.downsampley
@@ -65,6 +80,9 @@
     inputY = torch.tensor(inputY, dtype=torch.float)
     input = torch.stack([inputX, inputY], dim=-1)
 
+    ntrain =1000 
+    ntest = 200
+    print(ntrain)
     output = np.load(OUTPUT_Sigma)[:, 4]
     output = torch.tensor(output, dtype=torch.float)
     print(input.shape, output.shape)
@@ -73,45 +91,55 @@
     y_train = output[:ntrain, ::r1, ::r2][:, :s1, :s2]
     x_test = input[ntrain:ntrain + ntest, ::r1, ::r2][:, :s1, :s2]
     y_test = output[ntrain:ntrain + ntest, ::r1, ::r2][:, :s1, :s2]
+
     x_train = x_train.reshape(ntrain, -1, 2)
     x_test = x_test.reshape(ntest, -1, 2)
     y_train = y_train.reshape(ntrain, -1)
     y_test = y_test.reshape(ntest, -1)
 
-    train_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(x_train, x_train, y_train),
+    train_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(
+        x_train, x_train, y_train),
                                                batch_size=args.batch_size,
                                                shuffle=True)
-    test_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(x_test, x_test, y_test),
-                                              batch_size=args.batch_size,
+    test_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(
+        x_test, x_test, y_test),
+                                              batch_size=1,
                                               shuffle=False)
 
     print("Dataloading is over.")
-
-    model = get_model(args).Model(space_dim=2,
-                                  n_layers=args.n_layers,
-                                  n_hidden=args.n_hidden,
-                                  dropout=args.dropout,
-                                  n_head=args.n_heads,
-                                  Time_Input=False,
-                                  mlp_ratio=args.mlp_ratio,
-                                  fun_dim=0,
-                                  out_dim=1,
-                                  slice_num=args.slice_num,
-                                  ref=args.ref,
-                                  unified_pos=args.unified_pos,
-                                  H=s1, W=s2).cuda()
-
-    optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)
+    print(x_train.shape,x_test.shape)
+    model = LinearAttentionNeuralOperator(space_dim=2,
+                  n_layers=args.n_layers,
+                  n_hidden=args.n_hidden,
+                  dropout=args.dropout,
+                  n_head=args.n_heads,
+                  Time_Input=False,
+                  mlp_ratio=args.mlp_ratio,
+                  fun_dim=0,
+                  out_dim=1,
+                  key_ratio=args.key_ratio,
+                  ref=args.ref,
+                  unified_pos=args.unified_pos,
+                  H=s1,
+                  W=s2,
+                  args=args).cuda()
     print(args)
     print(model)
+    count_parameters1(model.blocks[0].Attn)
     count_parameters(model)
-
-    scheduler = torch.optim.lr_scheduler.OneCycleLR(optimizer, max_lr=args.lr, epochs=args.epochs,
-                                                    steps_per_epoch=len(train_loader))
+    optimizer = torch.optim.AdamW(model.parameters(),
+                                  lr=args.lr,
+                                  weight_decay=args.weight_decay)
+    scheduler = torch.optim.lr_scheduler.OneCycleLR(
+        optimizer,
+        max_lr=args.lr,
+        epochs=args.epochs,
+        steps_per_epoch=len(train_loader))
+    #l2 LOSS
     myloss = TestLoss(size_average=False)
 
     if eval:
-        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"))
+        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"),strict=True)
         model.eval()
         if not os.path.exists('./results/' + save_name + '/'):
             os.makedirs('./results/' + save_name + '/')
@@ -131,100 +159,140 @@
                 if id < showcase:
                     print(id)
                     plt.axis('off')
-                    plt.pcolormesh(x[0, :, 0].reshape(221, 51)[40:180, :35].detach().cpu().numpy(),
-                                   x[0, :, 1].reshape(221, 51)[40:180, :35].detach().cpu().numpy(),
-                                   np.zeros([140, 35]),
-                                   shading='auto',
-                                   edgecolors='black', linewidths=0.1)
-                    plt.colorbar()
-                    plt.savefig(
-                        os.path.join('./results/' + save_name + '/',
-                                     "input_" + str(id) + ".pdf"), bbox_inches='tight', pad_inches=0)
-                    plt.close()
-                    plt.axis('off')
-                    plt.pcolormesh(x[0, :, 0].reshape(221, 51)[40:180, :35].detach().cpu().numpy(),
-                                   x[0, :, 1].reshape(221, 51)[40:180, :35].detach().cpu().numpy(),
-                                   out[0, :].reshape(221, 51)[40:180, :35].detach().cpu().numpy(),
-                                   shading='auto', cmap='coolwarm')
+                    plt.pcolormesh(
+                        x[0, :,
+                          0].reshape(221,
+                                     51)[40:180, :35].detach().cpu().numpy(),
+                        x[0, :,
+                          1].reshape(221,
+                                     51)[40:180, :35].detach().cpu().numpy(),
+                        np.zeros([140, 35]),
+                        shading='auto',
+                        edgecolors='black',
+                        linewidths=0.1)
+                    plt.colorbar()
+                    plt.savefig(os.path.join('./results/' + save_name + '/',
+                                             "input_" + str(id) + ".pdf"),
+                                bbox_inches='tight',
+                                pad_inches=0)
+                    plt.close()
+                    plt.axis('off')
+                    plt.pcolormesh(
+                        x[0, :,
+                          0].reshape(221,
+                                     51)[40:180, :35].detach().cpu().numpy(),
+                        x[0, :,
+                          1].reshape(221,
+                                     51)[40:180, :35].detach().cpu().numpy(),
+                        out[0, :].reshape(
+                            221, 51)[40:180, :35].detach().cpu().numpy(),
+                        shading='auto',
+                        cmap='coolwarm')
                     plt.colorbar()
                     plt.clim(0, 1.2)
-                    plt.savefig(
-                        os.path.join('./results/' + save_name + '/',
-                                     "pred_" + str(id) + ".pdf"), bbox_inches='tight', pad_inches=0)
-                    plt.close()
-                    plt.axis('off')
-                    plt.pcolormesh(x[0, :, 0].reshape(221, 51)[40:180, :35].detach().cpu().numpy(),
-                                   x[0, :, 1].reshape(221, 51)[40:180, :35].detach().cpu().numpy(),
-                                   y[0, :].reshape(221, 51)[40:180, :35].detach().cpu().numpy(),
-                                   shading='auto', cmap='coolwarm')
+                    plt.savefig(os.path.join('./results/' + save_name + '/',
+                                             "pred_" + str(id) + ".pdf"),
+                                bbox_inches='tight',
+                                pad_inches=0)
+                    plt.close()
+                    plt.axis('off')
+                    plt.pcolormesh(
+                        x[0, :,
+                          0].reshape(221,
+                                     51)[40:180, :35].detach().cpu().numpy(),
+                        x[0, :,
+                          1].reshape(221,
+                                     51)[40:180, :35].detach().cpu().numpy(),
+                        y[0, :].reshape(
+                            221, 51)[40:180, :35].detach().cpu().numpy(),
+                        shading='auto',
+                        cmap='coolwarm')
                     plt.colorbar()
                     plt.clim(0, 1.2)
-                    plt.savefig(
-                        os.path.join('./results/' + save_name + '/',
-                                     "gt_" + str(id) + ".pdf"), bbox_inches='tight', pad_inches=0)
-                    plt.close()
-                    plt.axis('off')
-                    plt.pcolormesh(x[0, :, 0].reshape(221, 51)[40:180, :35].detach().cpu().numpy(),
-                                   x[0, :, 1].reshape(221, 51)[40:180, :35].detach().cpu().numpy(),
-                                   out[0, :].reshape(221, 51)[40:180, :35].detach().cpu().numpy() - \
-                                   y[0, :].reshape(221, 51)[40:180, :35].detach().cpu().numpy(),
-                                   shading='auto', cmap='coolwarm')
+                    plt.savefig(os.path.join('./results/' + save_name + '/',
+                                             "gt_" + str(id) + ".pdf"),
+                                bbox_inches='tight',
+                                pad_inches=0)
+                    plt.close()
+                    plt.axis('off')
+                    plt.pcolormesh(
+                        x[0, :,
+                          0].reshape(221,
+                                     51)[40:180, :35].detach().cpu().numpy(),
+                        x[0, :,
+                          1].reshape(221,
+                                     51)[40:180, :35].detach().cpu().numpy(),
+                        out[0, :].reshape(
+                            221, 51)[40:180, :35].detach().cpu().numpy() -
+                        y[0, :].reshape(
+                            221, 51)[40:180, :35].detach().cpu().numpy(),
+                        shading='auto',
+                        cmap='coolwarm')
                     plt.colorbar()
                     plt.clim(-0.2, 0.2)
-                    plt.savefig(
-                        os.path.join('./results/' + save_name + '/',
-                                     "error_" + str(id) + ".pdf"), bbox_inches='tight', pad_inches=0)
+                    plt.savefig(os.path.join('./results/' + save_name + '/',
+                                             "error_" + str(id) + ".pdf"),
+                                bbox_inches='tight',
+                                pad_inches=0)
                     plt.close()
 
         rel_err /= ntest
         print("rel_err:{}".format(rel_err))
+
     else:
-        for ep in range(args.epochs):
+        writer = SummaryWriter(log_dir='./runs/' + args.save_name)
+        for ep in tqdm(range(args.epochs)):
 
             model.train()
             train_loss = 0
 
-            for pos, fx, y in train_loader:
-
-                x, fx, y = pos.cuda(), fx.cuda(), y.cuda()  # x:B,N,2  fx:B,N,2  y:B,N
+            for pos, fx, y in tqdm(train_loader):
+
+                x, fx, y = pos.cuda(), fx.cuda(), y.cuda(
+                )  # x:B,N,2  fx:B,N,2  y:B,N
                 optimizer.zero_grad()
                 out = model(x, None).squeeze(-1)
+                # print(out.shape, y.shape)
                 loss = myloss(out, y)
                 loss.backward()
 
                 if args.max_grad_norm is not None:
-                    torch.nn.utils.clip_grad_norm_(model.parameters(), args.max_grad_norm)
+                    torch.nn.utils.clip_grad_norm_(model.parameters(),
+                                                   args.max_grad_norm)
                 optimizer.step()
                 train_loss += loss.item()
                 scheduler.step()
 
             train_loss = train_loss / ntrain
-            print("Epoch {} Train loss : {:.5f}".format(ep, train_loss))
-
-            model.eval()
-            rel_err = 0.0
-            with torch.no_grad():
-                for pos, fx, y in test_loader:
-                    x, fx, y = pos.cuda(), fx.cuda(), y.cuda()
-                    out = model(x, None).squeeze(-1)
-
-                    tl = myloss(out, y).item()
-                    rel_err += tl
-
-            rel_err /= ntest
-            print("rel_err:{}".format(rel_err))
-
-            if ep % 100 == 0:
+            tqdm.write("Epoch {} Train loss : {:.5f}".format(ep, train_loss))
+            writer.add_scalars('loss', {"train_loss": train_loss}, ep)
+
+            if (ep) % 10 == 0 or ep == args.epochs - 1:
+                model.eval()
+                rel_err = 0.0
+                with torch.no_grad():
+                    for pos, fx, y in test_loader:
+                        x, fx, y = pos.cuda(), fx.cuda(), y.cuda()
+                        out = model(x, None).squeeze(-1)
+
+                        tl = myloss(out, y).item()
+                        rel_err += tl
+
+                rel_err /= ntest
+                tqdm.write("rel_err:{}".format(rel_err))
+                writer.add_scalars('loss', {"rel_err": rel_err}, ep)
+
+            if (ep + 1) % 100 == 0:
                 if not os.path.exists('./checkpoints'):
                     os.makedirs('./checkpoints')
-                print('save model')
-                torch.save(model.state_dict(), os.path.join('./checkpoints', save_name + '.pt'))
+                tqdm.write('save model')
+                torch.save(model.state_dict(),
+                           os.path.join('./checkpoints', save_name + '.pt'))
 
         if not os.path.exists('./checkpoints'):
             os.makedirs('./checkpoints')
         print('save model')
-        torch.save(model.state_dict(), os.path.join('./checkpoints', save_name + '.pt'))
-
-
+        torch.save(model.state_dict(),
+                   os.path.join('./checkpoints', save_name + '.pt'))
 if __name__ == "__main__":
     main()
--- transolver/PDE-Solving-StandardBenchmark/exp_darcy.py
+++ linearno/Standard_PDE_Benchmark/exp_darcy.py
@@ -7,32 +7,33 @@
 from tqdm import *
 from utils.testloss import TestLoss
 from einops import rearrange
-from model_dict import get_model
 from utils.normalizer import UnitTransformer
 import matplotlib.pyplot as plt
-
-parser = argparse.ArgumentParser('Training Transolver')
+from torch.utils.tensorboard import SummaryWriter
+from model.LinearAttnNeuralOperator import LinearAttentionNeuralOperator
+
+parser = argparse.ArgumentParser('Training')
 
 parser.add_argument('--lr', type=float, default=1e-3)
 parser.add_argument('--epochs', type=int, default=500)
 parser.add_argument('--weight_decay', type=float, default=1e-5)
-parser.add_argument('--model', type=str, default='Transolver_2D')
 parser.add_argument('--n-hidden', type=int, default=64, help='hidden dim')
 parser.add_argument('--n-layers', type=int, default=3, help='layers')
 parser.add_argument('--n-heads', type=int, default=4)
 parser.add_argument('--batch-size', type=int, default=8)
-parser.add_argument("--gpu", type=str, default='1', help="GPU index to use")
+parser.add_argument("--gpu", type=str, default='0', help="GPU index to use")
 parser.add_argument('--max_grad_norm', type=float, default=None)
 parser.add_argument('--downsample', type=int, default=5)
+parser.add_argument('--ntrain', type=int, default=1000)
 parser.add_argument('--mlp_ratio', type=int, default=1)
 parser.add_argument('--dropout', type=float, default=0.0)
-parser.add_argument('--ntrain', type=int, default=1000)
 parser.add_argument('--unified_pos', type=int, default=0)
 parser.add_argument('--ref', type=int, default=8)
-parser.add_argument('--slice_num', type=int, default=32)
+parser.add_argument('--key_ratio', type=int, default=4)
 parser.add_argument('--eval', type=int, default=0)
-parser.add_argument('--save_name', type=str, default='darcy_Transolver')
+parser.add_argument('--save_name', type=str, default='darcy_LinearAttnNeuralOperator')
 parser.add_argument('--data_path', type=str, default='/data/fno')
+parser.add_argument("--model",type=str,default='None')
 args = parser.parse_args()
 
 os.environ["CUDA_VISIBLE_DEVICES"] = args.gpu
@@ -41,10 +42,14 @@
 test_path = args.data_path + '/piececonst_r421_N1024_smooth2.mat'
 ntrain = args.ntrain
 ntest = 200
-epochs = args.epochs
+epochs = 500
 eval = args.eval
 save_name = args.save_name
-
+def set_seed(num):
+    torch.manual_seed(num)
+    torch.cuda.manual_seed_all(num)
+    np.random.seed(num)
+    torch.backends.cudnn.deterministic = True
 
 def count_parameters(model):
     total_params = 0
@@ -73,7 +78,6 @@
     h = int(((421 - 1) / r) + 1)
     s = h
     dx = 1.0 / s
-
     train_data = scio.loadmat(train_path)
     x_train = train_data['coeff'][:ntrain, ::r, ::r][:, :s, :s]
     x_train = x_train.reshape(ntrain, -1)
@@ -109,34 +113,42 @@
     pos_train = pos.repeat(ntrain, 1, 1)
     pos_test = pos.repeat(ntest, 1, 1)
     print("Dataloading is over.")
-
+    print(x_train.shape, x_test.shape)
     train_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(pos_train, x_train, y_train),
                                                batch_size=args.batch_size, shuffle=True)
-    test_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(pos_test, x_test, y_test),
-                                              batch_size=args.batch_size, shuffle=False)
-
-    model = get_model(args).Model(space_dim=2,
-                                  n_layers=args.n_layers,
-                                  n_hidden=args.n_hidden,
-                                  dropout=args.dropout,
-                                  n_head=args.n_heads,
-                                  Time_Input=False,
-                                  mlp_ratio=args.mlp_ratio,
-                                  fun_dim=1,
-                                  out_dim=1,
-                                  slice_num=args.slice_num,
-                                  ref=args.ref,
-                                  unified_pos=args.unified_pos,
-                                  H=s, W=s).cuda()
-
-    optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)
+    test_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(
+        pos_test, x_test, y_test),
+                                              batch_size=1,
+                                              shuffle=False)
+    # if not args.is_mamba:
+    model = LinearAttentionNeuralOperator(space_dim=2,
+                     n_layers=args.n_layers,
+                     n_hidden=args.n_hidden,
+                     dropout=args.dropout,
+                     n_head=args.n_heads,
+                     Time_Input=False,
+                     mlp_ratio=args.mlp_ratio,
+                     fun_dim=1,
+                     out_dim=1,
+                     key_ratio=args.key_ratio,
+                     ref=args.ref,
+                     unified_pos=args.unified_pos,
+                     H=s,
+                     W=s,
+                     args=args).cuda()
+    optimizer = torch.optim.AdamW(model.parameters(),
+                                  lr=args.lr,
+                                  weight_decay=args.weight_decay)
 
     print(args)
     print(model)
     count_parameters(model)
-
-    scheduler = torch.optim.lr_scheduler.OneCycleLR(optimizer, max_lr=args.lr, epochs=epochs,
-                                                    steps_per_epoch=len(train_loader))
+    scheduler = torch.optim.lr_scheduler.OneCycleLR(
+        optimizer,
+        max_lr=args.lr,
+        epochs=epochs,
+        steps_per_epoch=len(train_loader))
+
     myloss = TestLoss(size_average=False)
     de_x = TestLoss(size_average=False)
     de_y = TestLoss(size_average=False)
@@ -144,9 +156,10 @@
     if eval:
         print("model evaluation")
         print(s, s)
-        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"), strict=False)
+        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"),
+                              strict=True)
         model.eval()
-        showcase = 10
+        showcase = 30
         id = 0
         if not os.path.exists('./results/' + save_name + '/'):
             os.makedirs('./results/' + save_name + '/')
@@ -164,10 +177,11 @@
                     rel_err += tl
 
                     if id < showcase:
-                        print(id)
-                        plt.figure()
-                        plt.axis('off')
-                        plt.imshow(out[0, :].reshape(85, 85).detach().cpu().numpy(), cmap='coolwarm')
+                        plt.figure()
+                        plt.axis('off')
+                        plt.imshow(out[0, :].reshape(
+                            85, 85).detach().cpu().numpy(),
+                                   cmap='coolwarm')
                         plt.colorbar()
                         plt.savefig(
                             os.path.join('./results/' + save_name + '/',
@@ -176,41 +190,51 @@
                         # ============ #
                         plt.figure()
                         plt.axis('off')
-                        plt.imshow(y[0, :].reshape(85, 85).detach().cpu().numpy(), cmap='coolwarm')
-                        plt.colorbar()
-                        plt.savefig(
-                            os.path.join('./results/' + save_name + '/', "case_" + str(id) + "_gt.pdf"))
+                        plt.imshow(y[0, :].reshape(85,
+                                                   85).detach().cpu().numpy(),
+                                   cmap='coolwarm')
+                        plt.colorbar()
+                        plt.savefig(
+                            os.path.join('./results/' + save_name + '/',
+                                         "case_" + str(id) + "_gt.pdf"))
                         plt.close()
                         # ============ #
                         plt.figure()
                         plt.axis('off')
-                        plt.imshow((y[0, :] - out[0, :]).reshape(85, 85).detach().cpu().numpy(), cmap='coolwarm')
+                        plt.imshow((y[0, :] - out[0, :]).reshape(
+                            85, 85).detach().cpu().numpy(),
+                                   cmap='coolwarm')
                         plt.colorbar()
                         plt.clim(-0.0005, 0.0005)
                         plt.savefig(
-                            os.path.join('./results/' + save_name + '/', "case_" + str(id) + "_error.pdf"))
+                            os.path.join('./results/' + save_name + '/',
+                                         "case_" + str(id) + "_error.pdf"))
                         plt.close()
                         # ============ #
                         plt.figure()
                         plt.axis('off')
-                        plt.imshow((fx[0, :].unsqueeze(-1)).reshape(85, 85).detach().cpu().numpy(), cmap='coolwarm')
-                        plt.colorbar()
-                        plt.savefig(
-                            os.path.join('./results/' + save_name + '/', "case_" + str(id) + "_input.pdf"))
+                        plt.imshow((fx[0, :].unsqueeze(-1)).reshape(
+                            85, 85).detach().cpu().numpy(),
+                                   cmap='coolwarm')
+                        plt.colorbar()
+                        plt.savefig(
+                            os.path.join('./results/' + save_name + '/',
+                                         "case_" + str(id) + "_input.pdf"))
                         plt.close()
 
             rel_err /= ntest
             print("rel_err:{}".format(rel_err))
     else:
-        for ep in range(args.epochs):
+        for ep in tqdm(range(args.epochs)):
             model.train()
             train_loss = 0
             reg = 0
-            for x, fx, y in train_loader:
+            for x, fx, y in tqdm(train_loader):
                 x, fx, y = x.cuda(), fx.cuda(), y.cuda()
                 optimizer.zero_grad()
 
-                out = model(x, fx=fx.unsqueeze(-1)).squeeze(-1)  # B, N , 2, fx: B, N, y: B, N
+                out = model(x, fx=fx.unsqueeze(-1)).squeeze(
+                    -1)  # B, N , 2, fx: B, N, y: B, N
                 out = y_normalizer.decode(out)
                 y = y_normalizer.decode(y)
 
@@ -222,12 +246,14 @@
                 out = rearrange(out, 'b c h w -> b (h w) c')
                 gt_grad_x, gt_grad_y = central_diff(y.unsqueeze(-1), dx, s)
                 pred_grad_x, pred_grad_y = central_diff(out, dx, s)
-                deriv_loss = de_x(pred_grad_x, gt_grad_x) + de_y(pred_grad_y, gt_grad_y)
+                deriv_loss = de_x(pred_grad_x, gt_grad_x) + de_y(
+                    pred_grad_y, gt_grad_y)
                 loss = 0.1 * deriv_loss + l2loss
                 loss.backward()
 
                 if args.max_grad_norm is not None:
-                    torch.nn.utils.clip_grad_norm_(model.parameters(), args.max_grad_norm)
+                    torch.nn.utils.clip_grad_norm_(model.parameters(),
+                                                   args.max_grad_norm)
                 optimizer.step()
                 train_loss += l2loss.item()
                 reg += deriv_loss.item()
@@ -235,7 +261,8 @@
 
             train_loss /= ntrain
             reg /= ntrain
-            print("Epoch {} Reg : {:.5f} Train loss : {:.5f}".format(ep, reg, train_loss))
+            tqdm.write("Epoch {} Reg : {:.5f} Train loss : {:.5f}".format(
+                ep, reg, train_loss))
 
             model.eval()
             rel_err = 0.0
@@ -243,10 +270,6 @@
             with torch.no_grad():
                 for x, fx, y in test_loader:
                     id += 1
-                    if id == 2:
-                        vis = True
-                    else:
-                        vis = False
                     x, fx, y = x.cuda(), fx.cuda(), y.cuda()
                     out = model(x, fx=fx.unsqueeze(-1)).squeeze(-1)
                     out = y_normalizer.decode(out)
@@ -254,19 +277,20 @@
                     rel_err += tl
 
             rel_err /= ntest
-            print("rel_err:{}".format(rel_err))
+            tqdm.write("rel_err:{}".format(rel_err))
 
             if ep % 100 == 0:
                 if not os.path.exists('./checkpoints'):
                     os.makedirs('./checkpoints')
-                print('save model')
-                torch.save(model.state_dict(), os.path.join('./checkpoints', save_name + '.pt'))
+                tqdm.write('save model')
+                torch.save(model.state_dict(),
+                           os.path.join('./checkpoints', save_name + '.pt'))
 
         if not os.path.exists('./checkpoints'):
             os.makedirs('./checkpoints')
         print('save model')
-        torch.save(model.state_dict(), os.path.join('./checkpoints', save_name + '.pt'))
-
+        torch.save(model.state_dict(),
+                   os.path.join('./checkpoints', save_name + '.pt'))
 
 if __name__ == "__main__":
     main()
--- transolver/PDE-Solving-StandardBenchmark/exp_elas.py
+++ linearno/Standard_PDE_Benchmark/exp_elas.py
@@ -5,38 +5,43 @@
 import torch
 from tqdm import *
 from utils.testloss import TestLoss
-from model_dict import get_model
 from utils.normalizer import UnitTransformer
+from model.LinearAttnNeuralOperator import LinearAttentionNeuralOperator
+from tqdm import tqdm
 
 parser = argparse.ArgumentParser('Training Transformer')
 
 parser.add_argument('--lr', type=float, default=1e-3)
 parser.add_argument('--epochs', type=int, default=500)
 parser.add_argument('--weight_decay', type=float, default=1e-5)
-parser.add_argument('--model', type=str, default='Transolver_1D')
 parser.add_argument('--n-hidden', type=int, default=64, help='hidden dim')
 parser.add_argument('--n-layers', type=int, default=3, help='layers')
 parser.add_argument('--n-heads', type=int, default=4)
 parser.add_argument('--batch-size', type=int, default=8)
-parser.add_argument("--gpu", type=str, default='1', help="GPU index to use")
+parser.add_argument("--gpu", type=str, default='0', help="GPU index to use")
 parser.add_argument('--max_grad_norm', type=float, default=None)
-parser.add_argument('--downsample', type=int, default=5)
+parser.add_argument('--downsample', type=int, default=1)
+parser.add_argument('--ntrain', type=int, default=1000)
 parser.add_argument('--mlp_ratio', type=int, default=1)
 parser.add_argument('--dropout', type=float, default=0.0)
-parser.add_argument('--ntrain', type=int, default=1000)
 parser.add_argument('--unified_pos', type=int, default=0)
 parser.add_argument('--ref', type=int, default=8)
-parser.add_argument('--slice_num', type=int, default=32)
+parser.add_argument('--key_ratio', type=int, default=4)
 parser.add_argument('--eval', type=int, default=0)
-parser.add_argument('--save_name', type=str, default='elas_Transolver')
+parser.add_argument('--save_name', type=str, default='elas_LinearAttnNeuralOperator')
 parser.add_argument('--data_path', type=str, default='/data/fno')
+parser.add_argument("--model",type=str,default='2D_temp')
 args = parser.parse_args()
 eval = args.eval
 save_name = args.save_name
 
 os.environ["CUDA_VISIBLE_DEVICES"] = args.gpu
 
-
+def set_seed(num):
+    torch.manual_seed(num)
+    torch.cuda.manual_seed_all(num)
+    np.random.seed(num)
+    torch.backends.cudnn.deterministic = True
 def count_parameters(model):
     total_params = 0
     for name, parameter in model.named_parameters():
@@ -50,7 +55,7 @@
 def main():
     ntrain = args.ntrain
     ntest = 200
-
+    # set_seed(0)
     PATH_Sigma = args.data_path + '/elasticity/Meshes/Random_UnitCell_sigma_10.npy'
     PATH_XY = args.data_path + '/elasticity/Meshes/Random_UnitCell_XY_10.npy'
 
@@ -71,40 +76,46 @@
     train_s = y_normalizer.encode(train_s)
     y_normalizer.cuda()
 
-    train_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(train_xy, train_xy, train_s),
+    train_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(
+        train_xy, train_xy, train_s),
                                                batch_size=args.batch_size,
                                                shuffle=True)
-    test_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(test_xy, test_xy, test_s),
-                                              batch_size=args.batch_size,
+    test_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(
+        test_xy, test_xy, test_s),
+                                              batch_size=1,
                                               shuffle=False)
 
     print("Dataloading is over.")
-
-    model = get_model(args).Model(space_dim=2,
-                                  n_layers=args.n_layers,
-                                  n_hidden=args.n_hidden,
-                                  dropout=args.dropout,
-                                  n_head=args.n_heads,
-                                  Time_Input=False,
-                                  mlp_ratio=args.mlp_ratio,
-                                  fun_dim=0,
-                                  out_dim=1,
-                                  slice_num=args.slice_num,
-                                  ref=args.ref,
-                                  unified_pos=args.unified_pos).cuda()
-
-    optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)
+    print(train_xy.shape,test_xy.shape)
+    model = LinearAttentionNeuralOperator(space_dim=2,
+                    n_layers=args.n_layers,
+                    n_hidden=args.n_hidden,
+                    dropout=args.dropout,
+                    n_head=args.n_heads,
+                    Time_Input=False,
+                    mlp_ratio=args.mlp_ratio,
+                    fun_dim=0,
+                    out_dim=1,
+                    key_ratio=args.key_ratio,
+                    ref=args.ref,
+                    unified_pos=args.unified_pos,args=args).cuda()
+    
+    optimizer = torch.optim.AdamW(model.parameters(),
+                                  lr=args.lr,
+                                  weight_decay=args.weight_decay)
 
     print(args)
     print(model)
     count_parameters(model)
-    
-    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=args.epochs)
-    
+
+    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer,
+                                                           T_max=args.epochs)
+
+
     myloss = TestLoss(size_average=False)
 
     if eval:
-        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"))
+        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"),strict=True)
         model.eval()
         if not os.path.exists('./results/' + save_name + '/'):
             os.makedirs('./results/' + save_name + '/')
@@ -123,46 +134,57 @@
                 if id < showcase:
                     print(id)
                     plt.axis('off')
-                    plt.scatter(x=fx[0, :, 0].detach().cpu().numpy(), y=fx[0, :, 1].detach().cpu().numpy(),
-                                c=y[0, :].detach().cpu().numpy(), cmap='coolwarm')
+                    plt.scatter(x=fx[0, :, 0].detach().cpu().numpy(),
+                                y=fx[0, :, 1].detach().cpu().numpy(),
+                                c=y[0, :].detach().cpu().numpy(),
+                                cmap='coolwarm')
                     plt.colorbar()
                     plt.clim(0, 1000)
-                    plt.savefig(
-                        os.path.join('./results/' + save_name + '/',
-                                     "gt_" + str(id) + ".pdf"), bbox_inches='tight', pad_inches=0)
+                    plt.savefig(os.path.join('./results/' + save_name + '/',
+                                             "gt_" + str(id) + ".pdf"),
+                                bbox_inches='tight',
+                                pad_inches=0)
                     plt.close()
 
                     plt.axis('off')
-                    plt.scatter(x=fx[0, :, 0].detach().cpu().numpy(), y=fx[0, :, 1].detach().cpu().numpy(),
-                                c=out[0, :].detach().cpu().numpy(), cmap='coolwarm')
+                    plt.scatter(x=fx[0, :, 0].detach().cpu().numpy(),
+                                y=fx[0, :, 1].detach().cpu().numpy(),
+                                c=out[0, :].detach().cpu().numpy(),
+                                cmap='coolwarm')
                     plt.colorbar()
                     plt.clim(0, 1000)
-                    plt.savefig(
-                        os.path.join('./results/' + save_name + '/',
-                                     "pred_" + str(id) + ".pdf"), bbox_inches='tight', pad_inches=0)
+                    plt.savefig(os.path.join('./results/' + save_name + '/',
+                                             "pred_" + str(id) + ".pdf"),
+                                bbox_inches='tight',
+                                pad_inches=0)
                     plt.close()
 
                     plt.axis('off')
-                    plt.scatter(x=fx[0, :, 0].detach().cpu().numpy(), y=fx[0, :, 1].detach().cpu().numpy(),
-                                c=((y[0, :] - out[0, :])).detach().cpu().numpy(), cmap='coolwarm')
+                    plt.scatter(x=fx[0, :, 0].detach().cpu().numpy(),
+                                y=fx[0, :, 1].detach().cpu().numpy(),
+                                c=((y[0, :] -
+                                    out[0, :])).detach().cpu().numpy(),
+                                cmap='coolwarm')
                     plt.clim(-8, 8)
                     plt.colorbar()
-                    plt.savefig(
-                        os.path.join('./results/' + save_name + '/',
-                                     "error_" + str(id) + ".pdf"), bbox_inches='tight', pad_inches=0)
+                    plt.savefig(os.path.join('./results/' + save_name + '/',
+                                             "error_" + str(id) + ".pdf"),
+                                bbox_inches='tight',
+                                pad_inches=0)
                     plt.close()
 
         rel_err /= ntest
         print("rel_err : {}".format(rel_err))
     else:
-        for ep in range(args.epochs):
+        for ep in tqdm(range(args.epochs)):
 
             model.train()
             train_loss = 0
 
-            for pos, fx, y in train_loader:
-
-                x, fx, y = pos.cuda(), fx.cuda(), y.cuda()  # x:B,N,2  fx:B,N,2  y:B,N,
+            for pos, fx, y in tqdm(train_loader):
+
+                x, fx, y = pos.cuda(), fx.cuda(), y.cuda(
+                )  # x:B,N,2  fx:B,N,2  y:B,N,
                 optimizer.zero_grad()
                 out = model(x, None).squeeze(-1)
                 out = y_normalizer.decode(out)
@@ -171,13 +193,14 @@
                 loss.backward()
 
                 if args.max_grad_norm is not None:
-                    torch.nn.utils.clip_grad_norm_(model.parameters(), args.max_grad_norm)
+                    torch.nn.utils.clip_grad_norm_(model.parameters(),
+                                                   args.max_grad_norm)
                 optimizer.step()
                 train_loss += loss.item()
             scheduler.step()
 
             train_loss = train_loss / ntrain
-            print("Epoch {} Train loss : {:.5f}".format(ep, train_loss))
+            tqdm.write("Epoch {} Train loss : {:.5f}".format(ep, train_loss))
 
             model.eval()
             rel_err = 0.0
@@ -190,19 +213,20 @@
                     rel_err += tl
 
             rel_err /= ntest
-            print("rel_err : {}".format(rel_err))
+            tqdm.write("rel_err : {}".format(rel_err))
 
             if ep % 100 == 0:
                 if not os.path.exists('./checkpoints'):
                     os.makedirs('./checkpoints')
-                print('save model')
-                torch.save(model.state_dict(), os.path.join('./checkpoints', save_name + '.pt'))
+                tqdm.write('save model')
+                torch.save(model.state_dict(),
+                           os.path.join('./checkpoints', save_name + '.pt'))
 
         if not os.path.exists('./checkpoints'):
             os.makedirs('./checkpoints')
         print('save model')
-        torch.save(model.state_dict(), os.path.join('./checkpoints', save_name + '.pt'))
-
+        torch.save(model.state_dict(),
+                   os.path.join('./checkpoints', save_name + '.pt'))
 
 if __name__ == "__main__":
     main()
--- transolver/PDE-Solving-StandardBenchmark/exp_ns.py
+++ linearno/Standard_PDE_Benchmark/exp_ns.py
@@ -6,14 +6,12 @@
 import torch
 from tqdm import *
 from utils.testloss import TestLoss
-from model_dict import get_model
-
-parser = argparse.ArgumentParser('Training Transformer')
-
+
+from tqdm import tqdm
+parser = argparse.ArgumentParser('Training')
 parser.add_argument('--lr', type=float, default=1e-3)
 parser.add_argument('--epochs', type=int, default=500)
 parser.add_argument('--weight_decay', type=float, default=1e-5)
-parser.add_argument('--model', type=str, default='Transolver_2D')
 parser.add_argument('--n-hidden', type=int, default=64, help='hidden dim')
 parser.add_argument('--n-layers', type=int, default=3, help='layers')
 parser.add_argument('--n-heads', type=int, default=4)
@@ -25,16 +23,16 @@
 parser.add_argument('--dropout', type=float, default=0.0)
 parser.add_argument('--unified_pos', type=int, default=0)
 parser.add_argument('--ref', type=int, default=8)
-parser.add_argument('--slice_num', type=int, default=32)
+parser.add_argument('--key_ratio', type=int, default=64)
 parser.add_argument('--eval', type=int, default=0)
-parser.add_argument('--save_name', type=str, default='ns_2d_UniPDE')
+parser.add_argument('--save_name', type=str, default='ns_UniPDE_LinearAttnNeuralOperator')
 parser.add_argument('--data_path', type=str, default='/data/fno')
+parser.add_argument("--model",type=str,default='None')
 args = parser.parse_args()
 
 os.environ["CUDA_VISIBLE_DEVICES"] = args.gpu
 
-data_path = args.data_path + '/NavierStokes_V1e-5_N1200_T20/NavierStokes_V1e-5_N1200_T20.mat'
-# data_path = args.data_path + '/NavierStokes_V1e-5_N1200_T20.mat'
+data_path = args.data_path + '/NavierStokes_V1e-5_N1200_T20.mat'
 ntrain = 1000
 ntest = 200
 T_in = 10
@@ -42,7 +40,11 @@
 step = 1
 eval = args.eval
 save_name = args.save_name
-
+def set_seed(num):
+    torch.manual_seed(num)
+    torch.cuda.manual_seed_all(num)
+    np.random.seed(num)
+    torch.backends.cudnn.deterministic = True
 
 def count_parameters(model):
     total_params = 0
@@ -57,7 +59,7 @@
 def main():
     r = args.downsample
     h = int(((64 - 1) / r) + 1)
-
+    # set_seed(0)
     data = scio.loadmat(data_path)
     print(data['u'].shape)
     train_a = data['u'][:ntrain, ::r, ::r, :T_in][:, :h, :h, :]
@@ -85,23 +87,24 @@
     train_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(pos_train, train_a, train_u),
                                                batch_size=args.batch_size, shuffle=True)
     test_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(pos_test, test_a, test_u),
-                                              batch_size=args.batch_size, shuffle=False)
+                                              batch_size=1, shuffle=False)
 
     print("Dataloading is over.")
-
-    model = get_model(args).Model(space_dim=2,
-                                  n_layers=args.n_layers,
-                                  n_hidden=args.n_hidden,
-                                  dropout=args.dropout,
-                                  n_head=args.n_heads,
-                                  Time_Input=False,
-                                  mlp_ratio=args.mlp_ratio,
-                                  fun_dim=T_in,
-                                  out_dim=1,
-                                  slice_num=args.slice_num,
-                                  ref=args.ref,
-                                  unified_pos=args.unified_pos,
-                                  H=h, W=h).cuda()
+    print(train_a.shape,test_a.shape)
+    from model.LinearAttnNeuralOperator import LinearAttentionNeuralOperator
+    model = LinearAttentionNeuralOperator(space_dim=2,
+                    n_layers=args.n_layers,
+                    n_hidden=args.n_hidden,
+                    dropout=args.dropout,
+                    n_head=args.n_heads,
+                    Time_Input=False,
+                    mlp_ratio=args.mlp_ratio,
+                    fun_dim=T_in,
+                    out_dim=1,
+                    key_ratio=args.key_ratio,
+                    ref=args.ref,
+                    unified_pos=args.unified_pos,
+                    H=h, W=h,args=args).cuda()
 
     optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)
 
@@ -114,7 +117,7 @@
     myloss = TestLoss(size_average=False)
 
     if eval:
-        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"), strict=False)
+        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"), strict=True)
         model.eval()
         showcase = 10
         id = 0
@@ -170,13 +173,13 @@
                 test_l2_full += myloss(pred.reshape(bsz, -1), yy.reshape(bsz, -1)).item()
             print(test_l2_full / ntest)
     else:
-        for ep in range(args.epochs):
+        for ep in tqdm(range(args.epochs)):
 
             model.train()
             train_l2_step = 0
             train_l2_full = 0
 
-            for x, fx, yy in train_loader:
+            for x, fx, yy in tqdm(train_loader):
                 loss = 0
                 x, fx, yy = x.cuda(), fx.cuda(), yy.cuda()  # x: B,4096,2    fx: B,4096,T   y: B,4096,T
                 bsz = x.shape[0]
@@ -189,8 +192,9 @@
                         pred = im
                     else:
                         pred = torch.cat((pred, im), -1)
-                    fx = torch.cat((fx[..., step:], y), dim=-1)  # detach() & groundtruth
-
+                    fx = torch.cat((fx[..., step:], y), dim=-1)
+                    # fx = torch.cat((fx[..., :2],fx[...,step+2:], y), dim=-1)  # detach() & groundtruth
+                    
                 train_l2_step += loss.item()
                 train_l2_full += myloss(pred.reshape(bsz, -1), yy.reshape(bsz, -1)).item()
                 optimizer.zero_grad()
@@ -219,11 +223,11 @@
                         else:
                             pred = torch.cat((pred, im), -1)
                         fx = torch.cat((fx[..., step:], im), dim=-1)
-
+                        # fx = torch.cat((fx[..., :2],fx[...,step+2:], im), dim=-1)
                     test_l2_step += loss.item()
                     test_l2_full += myloss(pred.reshape(bsz, -1), yy.reshape(bsz, -1)).item()
 
-            print(
+            tqdm.write(
                 "Epoch {} , train_step_loss:{:.5f} , train_full_loss:{:.5f} , test_step_loss:{:.5f} , test_full_loss:{:.5f}".format(
                     ep, train_l2_step / ntrain / (T / step), train_l2_full / ntrain, test_l2_step / ntest / (T / step),
                         test_l2_full / ntest))
@@ -231,7 +235,7 @@
             if ep % 100 == 0:
                 if not os.path.exists('./checkpoints'):
                     os.makedirs('./checkpoints')
-                print('save model')
+                tqdm.write('save model')
                 torch.save(model.state_dict(), os.path.join('./checkpoints', save_name + '.pt'))
 
         if not os.path.exists('./checkpoints'):
--- transolver/PDE-Solving-StandardBenchmark/exp_pipe.py
+++ linearno/Standard_PDE_Benchmark/exp_pipe.py
@@ -7,7 +7,6 @@
 parser.add_argument('--lr', type=float, default=1e-3)
 parser.add_argument('--epochs', type=int, default=500)
 parser.add_argument('--weight_decay', type=float, default=1e-5)
-parser.add_argument('--model', type=str, default='Transolver_2D')
 parser.add_argument('--n-hidden', type=int, default=64, help='hidden dim')
 parser.add_argument('--n-layers', type=int, default=3, help='layers')
 parser.add_argument('--n-heads', type=int, default=4)
@@ -20,10 +19,11 @@
 parser.add_argument('--dropout', type=float, default=0.0)
 parser.add_argument('--unified_pos', type=int, default=0)
 parser.add_argument('--ref', type=int, default=8)
-parser.add_argument('--slice_num', type=int, default=32)
+parser.add_argument('--key_ratio', type=int, default=4)
 parser.add_argument('--eval', type=int, default=0)
 parser.add_argument('--save_name', type=str, default='pipe_UniPDE')
 parser.add_argument('--data_path', type=str, default='/data/fno/pipe')
+parser.add_argument("--model",type=str,default='None')
 args = parser.parse_args()
 eval = args.eval
 save_name = args.save_name
@@ -34,10 +34,15 @@
 import torch
 from tqdm import *
 from utils.testloss import TestLoss
-from model_dict import get_model
+from model.LinearAttnNeuralOperator import LinearAttentionNeuralOperator
+from tqdm import tqdm
 from utils.normalizer import UnitTransformer
 
-
+def set_seed(num):
+    torch.manual_seed(num)
+    torch.cuda.manual_seed_all(num)
+    np.random.seed(num)
+    torch.backends.cudnn.deterministic = True
 def count_parameters(model):
     total_params = 0
     for name, parameter in model.named_parameters():
@@ -53,10 +58,6 @@
     INPUT_Y = args.data_path + '/Pipe_Y.npy'
     OUTPUT_Sigma = args.data_path + '/Pipe_Q.npy'
 
-    ntrain = 1000
-    ntest = 200
-    N = 1200
-
     r1 = args.downsamplex
     r2 = args.downsampley
     s1 = int(((129 - 1) / r1) + 1)
@@ -67,14 +68,17 @@
     inputY = np.load(INPUT_Y)
     inputY = torch.tensor(inputY, dtype=torch.float)
     input = torch.stack([inputX, inputY], dim=-1)
+    
+    ntrain = 1000
+    ntest = 200
 
     output = np.load(OUTPUT_Sigma)[:, 0]
     output = torch.tensor(output, dtype=torch.float)
     print(input.shape, output.shape)
-    x_train = input[:N][:ntrain, ::r1, ::r2][:, :s1, :s2]
-    y_train = output[:N][:ntrain, ::r1, ::r2][:, :s1, :s2]
-    x_test = input[:N][-ntest:, ::r1, ::r2][:, :s1, :s2]
-    y_test = output[:N][-ntest:, ::r1, ::r2][:, :s1, :s2]
+    x_train = input[:ntrain, ::r1, ::r2][:, :s1, :s2]
+    y_train = output[:ntrain, ::r1, ::r2][:, :s1, :s2]
+    x_test = input[-ntest:, ::r1, ::r2][:, :s1, :s2]
+    y_test = output[-ntest:, ::r1, ::r2][:, :s1, :s2]
     x_train = x_train.reshape(ntrain, -1, 2)
     x_test = x_test.reshape(ntest, -1, 2)
     y_train = y_train.reshape(ntrain, -1)
@@ -94,24 +98,25 @@
                                                batch_size=args.batch_size,
                                                shuffle=True)
     test_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(x_test, x_test, y_test),
-                                              batch_size=args.batch_size,
+                                              batch_size=1,
                                               shuffle=False)
 
     print("Dataloading is over.")
-
-    model = get_model(args).Model(space_dim=2,
-                                  n_layers=args.n_layers,
-                                  n_hidden=args.n_hidden,
-                                  dropout=args.dropout,
-                                  n_head=args.n_heads,
-                                  Time_Input=False,
-                                  mlp_ratio=args.mlp_ratio,
-                                  fun_dim=0,
-                                  out_dim=1,
-                                  slice_num=args.slice_num,
-                                  ref=args.ref,
-                                  unified_pos=args.unified_pos,
-                                  H=s1, W=s2).cuda()
+    print(x_train.shape,x_test.shape)
+    model = LinearAttentionNeuralOperator(space_dim=2,
+                    n_layers=args.n_layers,
+                    n_hidden=args.n_hidden,
+                    dropout=args.dropout,
+                    n_head=args.n_heads,
+                    Time_Input=False,
+                    mlp_ratio=args.mlp_ratio,
+                    fun_dim=0,
+                    out_dim=1,
+                    key_ratio=args.key_ratio,
+                    ref=args.ref,
+                    unified_pos=args.unified_pos,
+                    H=s1, W=s2,args=args).cuda()
+    
 
     optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)
 
@@ -125,8 +130,7 @@
     myloss = TestLoss(size_average=False)
 
     if eval:
-        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"), strict=False)
-        torch.save(model.state_dict(), os.path.join('./checkpoints', save_name + '_resave' + '.pt'))
+        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"), strict=True)
         model.eval()
         if not os.path.exists('./results/' + save_name + '/'):
             os.makedirs('./results/' + save_name + '/')
@@ -141,7 +145,7 @@
                 x, fx, y = pos.cuda(), fx.cuda(), y.cuda()
                 out = model(x, None).squeeze(-1)
                 out = y_normalizer.decode(out)
-
+              
                 tl = myloss(out, y).item()
                 rel_err += tl
 
@@ -187,7 +191,7 @@
                                    y[0, :].reshape(129, 129).detach().cpu().numpy(),
                                    shading='auto', cmap='coolwarm')
                     plt.colorbar()
-                    plt.clim(-0.02, 0.02)
+                    plt.clim(-0.005, 0.005)
                     plt.savefig(
                         os.path.join('./results/' + save_name + '/',
                                      "error_" + str(id) + ".pdf"), bbox_inches='tight', pad_inches=0)
@@ -196,12 +200,12 @@
         rel_err /= ntest
         print("rel_err:{}".format(rel_err))
     else:
-        for ep in range(args.epochs):
+        for ep in tqdm(range(args.epochs)):
 
             model.train()
             train_loss = 0
 
-            for pos, fx, y in train_loader:
+            for pos, fx, y in tqdm(train_loader):
 
                 x, fx, y = pos.cuda(), fx.cuda(), y.cuda()  # x:B,N,2  fx:B,N,2  y:B,N
                 optimizer.zero_grad()
@@ -209,7 +213,7 @@
 
                 out = y_normalizer.decode(out)
                 y = y_normalizer.decode(y)
-
+                
                 loss = myloss(out, y)
                 loss.backward()
 
@@ -221,7 +225,7 @@
                 scheduler.step()
 
             train_loss = train_loss / ntrain
-            print("Epoch {} Train loss : {:.5f}".format(ep, train_loss))
+            tqdm.write("Epoch {} Train loss : {:.5f}".format(ep, train_loss))
 
             model.eval()
             rel_err = 0.0
@@ -230,17 +234,17 @@
                     x, fx, y = pos.cuda(), fx.cuda(), y.cuda()
                     out = model(x, None).squeeze(-1)
                     out = y_normalizer.decode(out)
-
+                    
                     tl = myloss(out, y).item()
                     rel_err += tl
 
             rel_err /= ntest
-            print("rel_err:{}".format(rel_err))
+            tqdm.write("rel_err:{}".format(rel_err))
 
             if ep % 100 == 0:
                 if not os.path.exists('./checkpoints'):
                     os.makedirs('./checkpoints')
-                print('save model')
+                tqdm.write('save model')
                 torch.save(model.state_dict(), os.path.join('./checkpoints', save_name + '.pt'))
 
         if not os.path.exists('./checkpoints'):
--- transolver/PDE-Solving-StandardBenchmark/exp_plas.py
+++ linearno/Standard_PDE_Benchmark/exp_plas.py
@@ -2,12 +2,11 @@
 import argparse
 import matplotlib.pyplot as plt
 
-parser = argparse.ArgumentParser('Training Transformer')
+parser = argparse.ArgumentParser('Training')
 
 parser.add_argument('--lr', type=float, default=1e-3)
 parser.add_argument('--epochs', type=int, default=500)
 parser.add_argument('--weight_decay', type=float, default=1e-5)
-parser.add_argument('--model', type=str, default='Transolver_2D')
 parser.add_argument('--n-hidden', type=int, default=64, help='hidden dim')
 parser.add_argument('--n-layers', type=int, default=3, help='layers')
 parser.add_argument('--n-heads', type=int, default=4)
@@ -20,10 +19,11 @@
 parser.add_argument('--dropout', type=float, default=0.0)
 parser.add_argument('--unified_pos', type=int, default=0)
 parser.add_argument('--ref', type=int, default=8)
-parser.add_argument('--slice_num', type=int, default=32)
+parser.add_argument('--key_ratio', type=int, default=4)
 parser.add_argument('--eval', type=int, default=0)
-parser.add_argument('--save_name', type=str, default='plas_Transolver')
+parser.add_argument('--save_name', type=str, default='plas_LinearAttnNeuralOperator')
 parser.add_argument('--data_path', type=str, default='/data/fno/plas_N987_T20.mat')
+parser.add_argument("--model",type=str,default='None')
 args = parser.parse_args()
 eval = args.eval
 save_name = args.save_name
@@ -34,10 +34,15 @@
 import torch
 from tqdm import *
 from utils.testloss import TestLoss
-from model_dict import get_model
+from model.LinearAttnNeuralOperator import LinearAttentionNeuralOperator
+from tqdm import tqdm
 from utils.normalizer import UnitTransformer
 
-
+def set_seed(num):
+    torch.manual_seed(num)
+    torch.cuda.manual_seed_all(num)
+    np.random.seed(num)
+    torch.backends.cudnn.deterministic = True
 def count_parameters(model):
     total_params = 0
     for name, parameter in model.named_parameters():
@@ -86,8 +91,7 @@
 
 
 def main():
-    DATA_PATH = args.data_path
-
+    DATA_PATH = args.data_path + "/plas_N987_T20"
     N = 987
     ntrain = 900
     ntest = 80
@@ -130,6 +134,7 @@
     pos_train = pos.repeat(ntrain, 1, 1)
     pos_test = pos.repeat(ntest, 1, 1)
     print("Dataloading is over.")
+    print(x_train.shape,y_test.shape)
 
     t = np.linspace(0, 1, T)
     t = torch.tensor(t, dtype=torch.float).unsqueeze(0)
@@ -139,21 +144,22 @@
     train_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(pos_train, t_train, x_train, y_train),
                                                batch_size=args.batch_size, shuffle=True, collate_fn=random_collate_fn)
     test_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(pos_test, t_test, x_test, y_test),
-                                              batch_size=args.batch_size, shuffle=False)
+                                              batch_size=1, shuffle=False)
 
     print("Dataloading is over.")
-    model = get_model(args).Model(space_dim=2,
-                                  n_hidden=args.n_hidden,
-                                  n_layers=args.n_layers,
-                                  Time_Input=True,
-                                  n_head=args.n_heads,
-                                  fun_dim=1,
-                                  out_dim=Deformation,
-                                  mlp_ratio=args.mlp_ratio,
-                                  slice_num=args.slice_num,
-                                  unified_pos=args.unified_pos,
-                                  H=s1,
-                                  W=s2).cuda()
+    model = LinearAttentionNeuralOperator(space_dim=2,
+                    n_hidden=args.n_hidden,
+                    n_layers=args.n_layers,
+                    Time_Input=True,
+                    n_head=args.n_heads,
+                    fun_dim=1,
+                    out_dim=Deformation,
+                    mlp_ratio=args.mlp_ratio,
+                    key_ratio=args.key_ratio,
+
+                    unified_pos=args.unified_pos,
+                    H=s1,
+                    W=s2,args=args).cuda()
 
     optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)
 
@@ -166,7 +172,7 @@
     myloss = TestLoss(size_average=False)
 
     if eval:
-        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"), strict=False)
+        model.load_state_dict(torch.load("./checkpoints/" + save_name + ".pt"), strict=True)
         model.eval()
         if not os.path.exists('./results/' + save_name + '/'):
             os.makedirs('./results/' + save_name + '/')
@@ -180,17 +186,19 @@
                 loss = 0
                 x, fx, tim, yy = x.cuda(), fx.cuda(), tim.cuda(), yy.cuda()
                 bsz = x.shape[0]
-
+                
                 for t in range(T):
                     y = yy[..., t:t + 1]
                     input_T = tim[:, t:t + 1].reshape(bsz, 1)
+                    
                     im = model(x, fx, T=input_T)
                     loss += myloss(im.reshape(bsz, -1), y.reshape(bsz, -1))
+                    
                     if t == 0:
                         pred = im.unsqueeze(-1)
                     else:
                         pred = torch.cat((pred, im.unsqueeze(-1)), -1)
-
+                
                 if id < showcase:
                     print(id)
                     truth = y[0].reshape(101, 31, 4).squeeze().detach().cpu().numpy()
@@ -230,12 +238,12 @@
 
         print("test_step_loss:{:.5f} , test_full_loss:{:.5f}".format(test_l2_step / ntest / T, test_l2_full / ntest))
     else:
-        for ep in range(args.epochs):
+        for ep in tqdm(range(args.epochs)):
 
             model.train()
             train_l2_step = 0
 
-            for x, tim, fx, yy in train_loader:
+            for x, tim, fx, yy in tqdm(train_loader):
                 x, fx, tim, yy = x.cuda(), fx.cuda(), tim.cuda(), yy.cuda()
                 bsz = x.shape[0]
 
@@ -276,14 +284,14 @@
                     test_l2_step += loss.item()
                     test_l2_full += myloss(pred.reshape(bsz, -1), yy.reshape(bsz, -1)).item()
 
-            print("Epoch {} , train_step_loss:{:.5f} , test_step_loss:{:.5f} , test_full_loss:{:.5f}".format(ep,
+            tqdm.write("Epoch {} , train_step_loss:{:.5f} , test_step_loss:{:.5f} , test_full_loss:{:.5f}".format(ep,
                                                                                                              train_l2_step / ntrain / T,
                                                                                                              test_l2_step / ntest / T,
                                                                                                              test_l2_full / ntest))
             if ep % 100 == 0:
                 if not os.path.exists('./checkpoints'):
                     os.makedirs('./checkpoints')
-                print('save model')
+                tqdm.write('save model')
                 torch.save(model.state_dict(), os.path.join('./checkpoints', save_name + '.pt'))
 
         if not os.path.exists('./checkpoints'):
@@ -292,5 +300,6 @@
         torch.save(model.state_dict(), os.path.join('./checkpoints', save_name + '.pt'))
 
 
+
 if __name__ == "__main__":
     main()
